design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,153 @@
+# Overview
+VM Cloning is the action of creating a new VM from an existing one that has the exact same settings. This includes
+VM disks, network definition, CPU topology and so on.
+
+## Motivation
+The ability to clone a VM is a generic virtualization operation which may be useful for several use-cases.
+
+For example, a system admin / user might want to base a new VM on an existing one. In such a case we can clone a VM and
+then change it according to our needs. In addition, a cluster admin might want to try different configurations
+(for example, to check whether other configuration provide better performance for his workload), in this case he can
+clone the existing VM, test it and decide whether to move the new settings.
+
+Naively copy/pasting a VM spec would not work in most cases therefore this serves as an automation for cloning safely.
+
+## Goals
+Have a mechanism to clone VMs. Original and cloned VM would have the same:
+* Disks
+* Network configuration
+* Volumes
+* CPU topology
+
+Also, provide an ability to fine-tune what's being cloned. For example, it is possible to clone or skip
+annotations / labels.  
+
+## Non Goals
+In first implementation iteration the following limitations exist:
+* Only offline cloning will be supported, meaning cloning a VM when it's stopped.
+* Cloned VM would be in the same namespace as the original one.
+* Host disks would not be supported.
+
+## Definition of Users
+Everyone who creates a VM might create a VM clone.
+
+## User Stories
+* As a user / admin, I want to be able to create a new VM that is based on an existing VM.
+* As a user / admin, I want to create a clone to a VM using virtctl.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+Design can be inspired by the way VMIs are being migrated. In order to migrate a VMI, a [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration)
+CRD is posted into the cluster. This CRD is very thin and basically triggers a migration with basic parameters. Another
+example is VM snapshots / restores.
+
+For VM cloning a similar approach can be taken: a new CRD would be introduced that would provide some basic parameters like:
+* The VM that needs to be cloned
+* A new name for the VM (could be optional - a new name will be generated)
+* New MAC address (optional: if not provided MAC would be erased)
+* Which labels / annotations etc to clone or skip.
```

Comment:
@iholder-redhat there are (at least) a few more things to consider:

- smbios serial numbers
- disk serial numbers
- vtpm (once it lands)
- disks(!) - what should happen to each during cloning?

In general: Cloning is effectively the process to take a VM with an identity, and regenerating (or stripping) all identity related parts from the API, at the same time, a snapshot of the disk needs to be taken at the time of cloning.

I'm missing the discussion around these elements in this design. Was this discussed elsewhere?

Reply:
> what's being used in UI (cloning is fully implemented there [1]).

The UI is * very * naive, it create a dataVolume to copy the vms PVCs, and then attach the new PVC's.

Reply:
Thanks @yaacov.
What about the re-generation of things like smbios serial, disk serial, ssh hostname etc? How exactly is it handled from the UI perspective? Are you familiar with the QEMU function I was speaking of?

@awels - appreciate if you can take a look

Reply:
> How exactly is it handled from the UI perspective?

It is not, the UI just create a copy.

>  Are you familiar with the QEMU function I was speaking of?

No, the UI can only interact with the kubevirt API, it does not interact with the launcher pods

Reply:
I think we need Alex's help here :)

In the meantime, pretty much all other logic (except for this magical QEMU function) is implemented here https://github.com/kubevirt/kubevirt/pull/7336. This PR is a draft and will stay that way until https://github.com/kubevirt/kubevirt/pull/7318 merges.

Feedback is much appreciated.

Reply:
> @iholder-redhat there are (at least) a few more things to consider:
> 
> * smbios serial numbers
> * disk serial numbers
> * vtpm (once it lands)
> * disks(!) - what should happen to each during cloning?
> 
> In general: Cloning is effectively the process to take a VM with an identity, and regenerating (or stripping) all identity related parts from the API, at the same time, a snapshot of the disk needs to be taken at the time of cloning.
> 
> I'm missing the discussion around these elements in this design. Was this discussed elsewhere?

Hey @fabiand,

I came back with some answers :smiley: 

> * smbios serial numbers

We can specify these in VMI's spec, but don't have to as the field is optional. If not set, libvirt will regenerate one for us. So all that's needed here is to delete the smbios serial from the new VM's spec.

> * disk serial numbers

Disk serial being the same for different VMs shouldn't be a problem, as I understood from @vladikr. They are only used as an identifier from within the VM, not from the outer world.

> * vtpm (once it lands)

Two options that come into mind:
1. To not migrate VMIs with vTPM at all (can maybe be a first step).
2. Have a condition in VMI that's indicating of vTPM use. To migrate such VMs the user would need to explicitly enable a feature gate or an option in KubevirtCR's spec.

Something along those lines. Maybe @jean-edouard can give us his thoughts here.

> * disks(!) - what should happen to each during cloning?

This is completely covered by the snapshot mechanism. In short: whenever we can we use CDI's smart-cloning, and in other situation we rely on volume snapshots.

Another thing to take care of is the MAC address that can simply be deleted if kube-mac-pool is available. Otherwise, there's an option in Clone's spec to provide a manual MAC address.

Reply:
Thanks @jean-edouard!
Indeed sounds like something we should tackle in the future :)